draft-normington-jsonpath-latest.xml

```diff
@@ -104,7 +107,75 @@
       </section>
     </section>
    <section numbered="true" toc="default">
-      <name>JSONPath Syntax and Semantics</name>         
+      <name>JSONPath Syntax and Semantics</name>
+         <section numbered="true" toc="default">
+            <name>Overview</name>
+            <t>A JSONPath is a string which matches zero or more nodes of any JSON document. A valid JSONPath conforms to the ABNF syntax defined by this document.
+            </t>
+         </section>
+         <section numbered="true" toc="default">
+            <name>Implementation</name>
+            <t>An implementation of this specification, from now on referred to simply as "an implementation", SHOULD takes two inputs, a JSONPath and a JSON document, and produce
+               a possibly empty list of nodes of the JSON document which match the JSONPath or an error (but not both).
+            </t>
+            <t>If there is no matching node and no error has occurred, an implementation MUST return an empty list of nodes.
+            </t>
+            <t>Syntax errors in the JSONPath SHOULD be detected before matching is attempted since they do not depend on the JSON document.
+               Therefore, an implementation SHOULD take a JSONPath and produce an optional syntax error and then,
+               if and only if an error was not produced, SHOULD take a JSON document and produce a list of matches or an error (but not both).
+            </t>
+            <t>Alternatively, an implementation MAY take a JSONPath and a JSON document and produce a list of matches or an optional error (but not both).
+            </t>
+            <t>For any implementation, if a syntactically invalid JSONPath is provided, the implementation MUST return an error.
+            </t>
+            <t>If a syntactially invalid JSON document is provided, an implementation MAY return an error or MAY return an empty list of matches.
```

Comment:
> MAY return an error or MAY return an empty list of matches

What is the rationale behind letting the implementation to decide?

Reply:
I had in mind implementations, like my own, with an API consisting of:
* a constructor function which takes a JSONPath string and returns a rich JSONPath type or a syntax error. This gets syntax errors out of the way "early" (i.e. before any JSON documents are encountered) and allows "compilation" of the JSONPath to be performed once before the compiled version is used, possibly many times, to search JSON documents.
* a method on the JSONPath type which takes a _parsed_ form of a JSON document and which returns the search results or an error.

But you've identified a mistake there. Such implementations will _never_ encounter a syntactically invalid JSON document, so they needn't worry about how to handle syntax errors.

So I'm now leaning towards a stricter requirement:
> If a syntactically invalid JSON document is provided, an implementation MUST return an error.

WDYT?

Reply:
Yes, I think that's better.

Reply:
Ok. I'm now wondering how strict to be about checking the JSON document syntax because it's not really the prime responsibility of a JSONPath implementation. In particular, I don't want to over-burden JSONPath implementations in languages which happen to have imperfect JSON parsing libraries. So how about we weaken the wording to:
> If a syntactically invalid JSON document is provided, any implementation SHOULD return an error.

?

Reply:
Well, in fact many implementations will work without JSON text itself, using the result of JSON parsing as a structure; or even without JSON at all (yaml, for example). XML has DOM, that allows to isolate document structure from text source (and syntax); JSON lacks such tool.

Maybe we should just define JSON data structure as a tree in terms of typed and (partially) ordered nodes, thus totally isolating JsonPath from JSON parsing issues; but for now SHOULD is fairly enough.